### PR TITLE
[ownership] Update sil-combine for OSSA.

### PR DESCRIPTION
**Pull request by gottesmm**

### CODE REVIEW

File: lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp [hunk 1]
```diff
@@ -589,11 +570,27 @@ visitCheckedCastAddrBranchInst(CheckedCastAddrBranchInst *CCABI) {
...
+    // the top of the block if val is an argument.
+    {
+      auto *valInsertPt = val->getDefiningInsertionPoint();
+      if (!valInsertPt)
+        return nullptr;
+      auto valInsertPtIter = valInsertPt->getIterator();
+      // If our value is defined by an instruction (not an argument), we want to
+      // insert the copy after that. Otherwise, we have an argument and we want
+      // to insert the copy right at the beginning of the block.
+      if (val->getDefiningInstruction())
```
Comment (gottesmm) [inline lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:586]: @atrick I may be able to chop off this part into a separate PR.

File: lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp [hunk 2]
```diff
@@ -120,27 +164,29 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
...
-          origEnd->getLoc(), newBorrow);
-      pass.notifyNewInstruction(endBorrow);
+  SingleValueInstruction *loadedVal = *loadInst;
+  if (auto qual = loadInst.getOwnershipQualifier()) {
+    if (*qual == LoadOwnershipQualifier::Copy) {
+      // Borrow the load-copied subelement, with precisely the same scope as
+      // the aggregate borrow.
+      assert(extract->getNumOperands() == 1);
+      auto *origBorrow = cast<BeginBorrowInst>(extract->getOperand(0));
+      auto *newBorrow = SILBuilderWithScope(origBorrow)
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:178]: This I also may be able to split off into another PR.

File: lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp [hunk 3]
```diff
@@ -371,20 +340,36 @@ visitUnconditionalCheckedCastAddrInst(UnconditionalCheckedCastAddrInst *UCCAI) {
...
+        valInsertPtIter = std::next(valInsertPtIter);
+      SILBuilderWithScope builder(valInsertPtIter, Builder);
```
Comment (gottesmm) [inline lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp:358]: This I might be able to split off into another PR.

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp [hunk 4]
```diff
@@ -1284,19 +1280,19 @@ SILCombiner::propagateConcreteTypeOfInitExistential(FullApplySite Apply) {
...
+    auto *user = use->getUser();
+    if (user == ignoreUser)
       continue;
 
-    if (!isa<RetainValueInst>(Use->getUser()) &&
-        !isa<ReleaseValueInst>(Use->getUser()) &&
-        !isa<StrongRetainInst>(Use->getUser()) &&
-        !isa<StrongReleaseInst>(Use->getUser()))
+    if (!isa<RetainValueInst>(user) && !isa<ReleaseValueInst>(user) &&
+        !isa<StrongRetainInst>(user) && !isa<StrongReleaseInst>(user) &&
```
Comment (gottesmm) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:1291]: I didn't need to do this for any test cases, but I could add begin_borrow here.

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp [hunk 5]
```diff
@@ -570,13 +566,13 @@ SILCombiner::recursivelyCollectARCUsers(UserListTy &Uses, ValueBase *Value) {
...
-    if (isa<RefCountingInst>(Inst) ||
-        isa<DebugValueInst>(Inst)) {
+    if (isa<RefCountingInst>(Inst) || isa<DestroyValueInst>(Inst) ||
+        isa<DebugValueInst>(Inst) || isa<EndBorrowInst>(Inst)) {
       Uses.push_back(Inst);
       continue;
     }
-    if (isa<TupleExtractInst>(Inst) ||
-        isa<StructExtractInst>(Inst) ||
+    if (isa<TupleExtractInst>(Inst) || isa<StructExtractInst>(Inst) ||
```
Comment (gottesmm) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:574]: I imagine we will need to expand this to support destructure as well.

File: lib/SILOptimizer/Analysis/SimplifyInstruction.cpp [hunk 6]
```diff
@@ -298,11 +327,41 @@ SILValue InstSimplifier::visitAddressToPointerInst(AddressToPointerInst *ATPI) {
...
+      // that rematerializes the address is outside the borrow scope and has
+      // uses outside of the borrow scope, we could introduce an ownership
+      // violation.
+      //
+      // TODO: We need to be able to exhaustively traverse all uses of the
+      // resulting address to determine if all are within the borrow scope. This
+      // requires using the access scope. Instead of doing that, we just bail
+      // here.
+      SILValue baseAddr =
+          getUnderlyingObjectStoppingAtMixedProjections(ATPI->getOperand());
```
Comment (gottesmm) [inline lib/SILOptimizer/Analysis/SimplifyInstruction.cpp:357]: @atrick this is the issue I was telling you about with respect to address to pointer <-> pointer to address. I just tried to recognize if we had an interior pointer operand and bail in that case.

**Review by gottesmm — COMMENTED** [review]
@atrick some initial guidance comments.

**Comment by atrick** [comment]
@gottesmm I'll hold off reviewing this for now because is may be out-of-date

**Comment by gottesmm** [comment]
This was done in a different PR and is stale. Closing!
